src/app/gps/qgsappgpslogging.cpp

```diff
@@ -0,0 +1,387 @@
+  {
+    if ( mGpkgLogger )
+      mGpkgLogger->setEllipsoid( QgsProject::instance()->ellipsoid() );
+  } );
+
+  connect( mConnection, &QgsAppGpsConnection::connected, this, &QgsAppGpsLogging::gpsConnected );
+  connect( mConnection, &QgsAppGpsConnection::disconnected, this, &QgsAppGpsLogging::gpsDisconnected );
+
+  connect( QgsGui::instance(), &QgsGui::optionsChanged, this, [ = ]
+  {
+    if ( mGpkgLogger )
+      mGpkgLogger->updateGpsSettings();
+  } );
+}
+
+QgsAppGpsLogging::~QgsAppGpsLogging()
+{
+  if ( mGpkgLogger )
+  {
+    mGpkgLogger->endCurrentTrack();
+  }
+}
+
+void QgsAppGpsLogging::setNmeaLogFile( const QString &filename )
+{
+  if ( mLogFile )
+  {
+    stopNmeaLogging();
+  }
+
+  mNmeaLogFile = filename;
+
+  if ( mEnableNmeaLogging && !mNmeaLogFile.isEmpty() )
+  {
+    startNmeaLogging();
+  }
+}
+
+void QgsAppGpsLogging::setNmeaLoggingEnabled( bool enabled )
+{
+  if ( enabled == static_cast< bool >( mLogFile ) )
+    return;
+
+  if ( mLogFile && !enabled )
+  {
+    stopNmeaLogging();
+  }
+
+  mEnableNmeaLogging = enabled;
+
+  if ( mEnableNmeaLogging && !mNmeaLogFile.isEmpty() )
+  {
+    startNmeaLogging();
+  }
+}
+
+void QgsAppGpsLogging::setGpkgLogFile( const QString &filename )
+{
+  mGpkgLogFile = filename;
+  if ( filename.isEmpty() )
+  {
+    // stop logging
+    if ( mGpkgLogger )
+    {
+      mGpkgLogger->endCurrentTrack();
+      mGpkgLogger.reset();
+
+      QgisApp::instance()->messageBar()->pushInfo( QString(), tr( "GPS logging stopped" ) );
+    }
+  }
+  else
+  {
+    if ( !createOrUpdateLogDatabase() )
+      return;
+    createGpkgLogger();
+
+    QgisApp::instance()->messageBar()->pushInfo( QString(), tr( "Saving GPS log to <a href=\"%1\">%2</a>" ).arg( QUrl::fromLocalFile( mGpkgLogFile ).toString(), QDir::toNativeSeparators( mGpkgLogFile ) ) );
+  }
+}
+
+void QgsAppGpsLogging::gpsConnected()
+{
+  if ( !mLogFile && mEnableNmeaLogging && !mNmeaLogFile.isEmpty() )
+  {
+    startNmeaLogging();
+  }
+  if ( mGpkgLogger )
+  {
+    mGpkgLogger->setConnection( mConnection->connection() );
+  }
+}
+
+void QgsAppGpsLogging::gpsDisconnected()
+{
+  stopNmeaLogging();
+  if ( mGpkgLogger )
+  {
+    mGpkgLogger->endCurrentTrack();
+    mGpkgLogger->setConnection( nullptr );
+  }
+}
+
+void QgsAppGpsLogging::logNmeaSentence( const QString &nmeaString )
+{
+  if ( mEnableNmeaLogging && mLogFile && mLogFile->isOpen() )
+  {
+    mLogFileTextStream << nmeaString << "\r\n"; // specifically output CR + LF (NMEA requirement)
+  }
+}
+
+void QgsAppGpsLogging::startNmeaLogging()
+{
+  if ( !mLogFile )
+  {
+    mLogFile = std::make_unique< QFile >( mNmeaLogFile );
+  }
+
+  if ( mLogFile->open( QIODevice::Append ) )  // open in binary and explicitly output CR + LF per NMEA
+  {
+    mLogFileTextStream.setDevice( mLogFile.get() );
+
+    // crude way to separate chunks - use when manually editing file - NMEA parsers should discard
+    mLogFileTextStream << "====" << "\r\n";
```

Comment:
Maybe only add a separator if ( mLogFile.size() > 0 ) ?

Reply:
Ok, done. It's rather old code, I'm not sure if that whole "crude way" is required anymore!